File: RequirementsAnalysis/requirement-agent/requirement-agent/specValidator.ts
import type { RequirementValidationError } from "./types";

const REQUIRED_TOP_LEVEL_FIELDS = [
  "basicInfo",
  "background",
  "goals",
  "impactScope",
  "uiux",
  "acceptanceCriteria",
  "performanceRequirements",
  "compatibilityRequirements",
  "copywriting",
  "risks",
  "definitionOfDone",
  "openQuestions",
];

const REQUIRED_FIELD_PATHS = [
  "basicInfo.requirementName",
  "basicInfo.requirementType",
  "basicInfo.priority",
  "basicInfo.status",
  "background.context",
  "goals.inScope",
  "goals.outOfScope",
  "acceptanceCriteria.checklist",
  "acceptanceCriteria.gherkinScenarios",
];

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasPath(value: unknown, path: string): boolean {
  const segments = path.split(".");
  let current: unknown = value;

  for (const segment of segments) {
    if (!isObject(current) || !(segment in current)) {
      return false;
    }

    current = current[segment];
  }

  return true;
}

function missingFieldError(path: string): RequirementValidationError {
  return {
    category: "schema",
    keyword: path,
    message: `LLM 返回内容缺少必要字段：${path}`,
  };
}

export function validateRequirementSpecShape(spec: unknown): {
  valid: boolean;
  errors: RequirementValidationError[];
} {
  const errors: RequirementValidationError[] = [];

  if (!isObject(spec)) {
    return {
      valid: false,
      errors: [missingFieldError("RequirementSpec")],
    };
  }

  for (const field of REQUIRED_TOP_LEVEL_FIELDS) {
    if (!hasPath(spec, field)) {
      errors.push(missingFieldError(field));
    }
  }

  for (const path of REQUIRED_FIELD_PATHS) {
    if (!hasPath(spec, path)) {
      errors.push(missingFieldError(path));
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

File: ReqAnalysis/requirement_analysis/spec_validator.py
from pydantic import ValidationError

from .models import RequirementSpec, RequirementValidationError


def validate_requirement_spec_shape(
    spec: object,
) -> tuple[bool, list[RequirementValidationError], RequirementSpec | None]:
    try:
        parsed = RequirementSpec.model_validate(spec)
    except ValidationError as err:
        validation_errors: list[RequirementValidationError] = []
        for item in err.errors():
            location = (
                ".".join(str(part) for part in item.get("loc", []))
                or "RequirementSpec"
            )
            validation_errors.append(
                RequirementValidationError(
                    category="schema",
                    keyword=location,
                    message=f"LLM 返回内容缺少必要字段或字段类型错误：{location}",
                )
            )
        return False, validation_errors, None

    errors: list[RequirementValidationError] = []
    if len(parsed.goals.inScope) == 0:
        errors.append(
            RequirementValidationError(
                category="schema",
                keyword="goals.inScope",
                message="LLM 返回内容缺少必要字段：goals.inScope",
            )
        )
    if len(parsed.acceptanceCriteria.checklist) == 0:
        errors.append(
            RequirementValidationError(
                category="schema",
                keyword="acceptanceCriteria.checklist",
                message="LLM 返回内容缺少必要字段：acceptanceCriteria.checklist",
            )
        )

    return len(errors) == 0, errors, parsed
